From 12c957f786b12a4dd116f9c40a715d671d17fa16 Mon Sep 17 00:00:00 2001
From: Allan McRae <allan@archlinux.org>
Date: Mon, 12 Mar 2012 09:28:01 +0100
Subject: [PATCH] grep: fix segfault with -r --exclude-dir and no file operand

* src/main.c (grepdir): Don't invoke excluded_file_name on NULL.
* NEWS (Bug fixes): Mention it.
---
 NEWS       |    5 +++++
 src/main.c |    2 +-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/NEWS b/NEWS
index d0a63d5..d4d70f5 100644
--- a/NEWS
+++ b/NEWS
@@ -2,6 +2,11 @@ GNU grep NEWS                                    -*- outline -*-
 
 * Noteworthy changes in release ?.? (????-??-??) [?]
 
+** Bug fixes
+
+   grep no longer segfaults with -r --exclude-dir and no file operand.
+   I.e., ":|grep -r --exclude-dir=D PAT" would segfault.
+
 
 * Noteworthy changes in release 2.11 (2012-03-02) [stable]
 
diff --git a/src/main.c b/src/main.c
index 2f6c761..f4f1235 100644
--- a/src/main.c
+++ b/src/main.c
@@ -1361,7 +1361,7 @@ grepdir (char const *dir, struct stats const *stats)
   struct stats const *ancestor;
   char *name_space;
   int status = 1;
-  if (excluded_directory_patterns
+  if (dir && excluded_directory_patterns
       && excluded_file_name (excluded_directory_patterns, dir))
     return 1;
 
-- 
1.7.9.3

